middle_end/flambda2/to_cmm/un_cps_helper.ml

```diff
@@ -165,11 +165,58 @@ let xor_ = binary Cmm.Cxor
 let eq ?(dbg=Debuginfo.none) x y =
   match x, y with
   | Cmm.Cconst_int (n, _), Cmm.Cop(Csubi, [Cconst_int (m, _); c], _)
-  | Cmm.Cop(Csubi, [Cconst_int (m, _); c], _), Cmm.Cconst_int (n, _) ->
+  | Cmm.Cop(Csubi, [Cconst_int (m, _); c], _), Cmm.Cconst_int (n, _)
+    when Misc.no_overflow_sub m n ->
     (* [n = m - c] <=> [c = m - n]
        This is typically generated by expressions of the form
        [if not expr then ...], with [not expr] being compiled to [4 - c]
-       and the condition for the test becomes [1 = 4 - c]. *)
+       and the condition for the test becomes [1 = 4 - c].
+
+       We need to impose the side condition because the above equivalence hides
+       a subtlety: While [c] is a full-blooded native integer, [m] and [n]
+       are OCaml ints that get sign-extended at run time. That in itself doesn't
```

Comment:
Nitpick: they don't get sign-extended at runtime, they get sign-extended later in the backend (during `Selection`). The problem is the same though.

Reply:
Ah, good point. Better to say they'll _have been_ sign-extended by then.